rad/src/lib.rs

```diff
@@ -26,56 +28,97 @@ pub mod types;
 
 pub type Result<T> = std::result::Result<T, RadError>;
 
-pub async fn try_data_request(request: &RADRequest) -> RadonReport<RadonTypes> {
+/// The return type of any method executing the entire life cycle of a data request.
+#[derive(Debug, Serialize)]
+pub struct RADRequestExecutionReport {
+    /// Vector of reports about retrieval of data sources.
+    pub retrieve: Vec<RadonReport<RadonTypes>>,
+    /// Report about aggregation of data sources.
+    pub aggregate: RadonReport<RadonTypes>,
+    /// Report about aggregation of reports (reveals, actually).
+    pub tally: RadonReport<RadonTypes>,
+}
+
+pub async fn try_data_request(
+    request: &RADRequest,
+    settings: RadonScriptExecutionSettings,
+    inputs_injection: Option<&[&str]>,
+) -> RADRequestExecutionReport {
     let context = &mut ReportContext::default();
 
-    let retrieve_responses_fut = request
-        .retrieve
-        .iter()
-        .map(|retrieve| run_retrieval_report(retrieve));
+    let mut retrieve_responses = Vec::new();
 
-    let retrieve_responses: Vec<RadonTypes> = futures::future::join_all(retrieve_responses_fut)
-        .await
+    if let Some(inputs) = inputs_injection {
+        for (retrieve, input) in request.retrieve.iter().zip(inputs.iter()) {
+            retrieve_responses.push(run_retrieval_with_data_report(
+                retrieve, input, context, settings,
+            ))
+        }
+    } else {
+        for retrieve in &request.retrieve {
+            retrieve_responses.push(run_retrieval_report(retrieve, settings).await)
+        }
+    }
+
+    let retrieval_reports: Vec<RadonReport<RadonTypes>> = retrieve_responses
         .into_iter()
         .map(|retrieve| {
-            retrieve
-                .unwrap_or_else(|error| RadonReport::from_result(Err(error), context))
-                .into_inner()
+            retrieve.unwrap_or_else(|error| RadonReport::from_result(Err(error), context))
         })
         .collect();
+    let retrieval_values: Vec<RadonTypes> = retrieval_reports
+        .clone()
+        .into_iter()
+        .map(|report| report.into_inner())
```

Comment:
By using a clone here you are creating a temporary vector that is not needed, always prefer `.iter().cloned()`, or in this case:

```suggestion
        .iter()
        .map(|report| report.result.clone())
```